// @civic/auth/dist/shared/components/BlockDisplay.js
import { jsx as _jsx } from "@emotion/react/jsx-runtime";
import {} from "react";
import { useIframe } from "../hooks/useIframe.js";
import { useIsClient } from "usehooks-ts";
const BlockDisplay = ({ children }) => {
    const { backgroundColor } = useIframe();
    const isClient = useIsClient();
    // don't render in server-mode as the appearance changes after load
    return !isClient ? null : (_jsx("div", { id: "iframe-block-display-wrapper", style: {
            position: "absolute",
            left: 0,
            top: 0,
            zIndex: 200,
            display: "flex",
            height: "100vh",
            width: "100vw",
            alignItems: "center",
            justifyContent: "center",
            backgroundColor,
        }, children: _jsx("div", { id: "iframe-block-display", style: {
                position: "relative",
                inset: 0,
                display: "flex",
                alignItems: "center",
                justifyContent: "center",
                backgroundColor,
            }, children: children }) }));
};
export { BlockDisplay };
