// metakeep/webpack.config.js
const path = require("path");
const webpack = require("webpack");
const Dotenv = require("dotenv-webpack");

module.exports = (env) => ({
  mode: "production",
  entry: path.resolve(__dirname, "./src/index.ts"),

  output: {
    path: path.resolve(__dirname, "lib"),
    // This is needed to fix `Error: Automatic publicPath is not supported in this browser`
    publicPath: "/",
    filename: "index.js",
    library: {
      type: "umd",
      umdNamedDefine: true,
    },
    globalObject: "this",

    // Disable chunking to prevent error:
    // "Cannot read properties of undefined (reading 'webpackChunkmetakeep')""
    chunkFormat: false,
  },

  module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: ["babel-loader"],
      },
      {
        test: /\.tsx?$/,
        use: "ts-loader",
        exclude: /node_modules/,
      },
    ],
  },
  resolve: {
    extensions: [".tsx", ".ts", ".js"],
    fallback: {
      crypto: false,
    },
  },
  plugins: [
    new Dotenv({
      path: env.production ? "./.env.prod" : "./.env.dev",
    }),
  ],
});
